File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    player = {name: i for i, name in enumerate(players)}
    
    for i in callings:
        win_idx = player[i]
        roser = players[win_idx-1]
        players[win_idx-1] = i
        player[i] = win_idx-1
        players[win_idx] = roser
        player[roser] = win_idx
        
    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        int n = players.length;
        
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(players[i], i);
        }
        
        for (String call : callings) {
            int idx = map.get(call);
            String tmp = players[idx-1];
            players[idx-1] = call;
            players[idx] = tmp;
            
            map.put(call, idx-1);
            map.put(tmp, idx);
        }
        
        return players;
    }
}